fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -505,16 +504,15 @@ def _sync_params_and_buffers(self) -> None:
         Sync the complete model states in between the ranks
         """
 
-        last_work_handle = None
+        work_handles = []
 
         for t in self.module.state_dict().values():
-            last_work_handle = dist.broadcast(
-                t, src=self.reference_global_rank, group=self.process_group, async_op=True
+            work_handles.append(
+                dist.broadcast(t, src=self.reference_global_rank, group=self.process_group, async_op=True)
             )
 
-        # Only wait for the last handle, they're inlined in the same CUDA stream
-        if last_work_handle:
-            last_work_handle.wait()
+        # gloo does not guarantee inlining like NCCL, wait for all requests
+        _ = list(filter(lambda x: x.wait(), work_handles))
```

Comment:
the crux of the diff: Gloo is actually not guaranteeing that the network calls will be sequential (or so I've been told), which means that it's safer to actually check for all handles (not needed for NCCL, but I figured that a backend check here would really be ugly). same below

Reply:
is there a perf difference for NCCL? I vaguely remember this change helped with perf but I could have misremembered the PR.

Reply:
yes there's a small perf hit, you can see all the checks on the trace which take a non-negligible amount of time (like, %), but I think that having something super robust is still worth it for now, until we have a real perf-constrained usecase ? right now the broadcast part is slowing things down so much that this is o()

Reply:
(done)